cmd/bake/main.go

```diff
@@ -104,81 +150,83 @@ func (m *Main) Run() error {
 	// Recursively attach all readers to stdout/stderr.
 	m.pipeReaders(build, make(map[*bake.Build]struct{}))
 
-	// Execute the build.
-	if err := m.build(build); err != nil {
+	// Create mount directory.
+	mountPath, err := ioutil.TempDir("", "bake-")
+	if err != nil {
 		return err
 	}
+	defer os.Remove(mountPath)
 
-	return nil
-}
-
-// build executes a build.
-func (m *Main) build(build *bake.Build) error {
-	// Create a temporary directory for mounting.
-	path, err := ioutil.TempDir("", "bake-")
+	// Create file system.
+	fs, err := m.openFileSystem(mountPath)
 	if err != nil {
-		return fmt.Errorf("temp dir: %s", err)
+		return err
 	}
-	defer os.Remove(path)
+	defer m.closeFileSystem(fs)
 
-	// Convert root to an absolute path.
-	root, err := filepath.Abs(m.Path)
-	if err != nil {
-		return fmt.Errorf("abs path: %s", err)
+	// Execute the build.
+	if err := m.build(build, fs, ss); err != nil {
+		return err
 	}
 
+	return nil
+}
+
+// openFileSystem initializes and mounts a file system to a temporary directory.
+func (m *Main) openFileSystem(mountPath string) (bake.FileSystem, error) {
 	// Create file system.
 	fs, err := bake.NewFileSystem(DefaultFileSystem, bake.FileSystemOptions{
-		Root: root,
+		Path:      m.Root,
+		MountPath: mountPath,
 	})
 	if err != nil {
-		return fmt.Errorf("new file system: %s", err)
+		return nil, fmt.Errorf("new file system: %s", err)
 	}
 
 	// Open file system.
 	if err := fs.Open(); err != nil {
-		return fmt.Errorf("open file system: %s", err)
+		return nil, fmt.Errorf("open file system: %s", err)
 	}
-	defer fs.Close()
 
-	// Mount file system to temporary directory.
-	if err := fs.Mount(path); err != nil {
-		return fmt.Errorf("mount: %s", err)
-	}
-	defer fs.Unmount(path)
+	return fs, nil
+}
 
-	// Save working directory to restore when program is done.
-	wd, err := os.Getwd()
-	if err != nil {
+// closeFileSystem shuts down and unmounts the file system.
+// Also removes underlying mount directory.
+func (m *Main) closeFileSystem(fs bake.FileSystem) error {
+	if err := fs.Close(); err != nil {
 		return err
 	}
-	defer os.Chdir(wd)
+	return nil
+}
 
-	// Change to new directory
-	if err := os.Chdir(path); err != nil {
-		return err
-	}
+// build executes a build against a file system.
+func (m *Main) build(build *bake.Build, fs bake.FileSystem, ss *bake.Snapshot) error {
+	/*
+		// Save working directory to restore when program is done.
+		wd, err := os.Getwd()
+		if err != nil {
+			return err
+		}
+		defer os.Chdir(wd)
+
+		// Change to new directory
+		if err := os.Chdir(path); err != nil {
+			return err
+		}
+	*/
```

Comment:
The pwd was moved to the `exec.Cmd` so it wasn't needed here. I forgot to pull out the commented section. Fixed in 9fae1c3.